qa/setup_packages.py

```diff
@@ -16,34 +16,433 @@
-                        "90" : ["1.6.0"],
-                        "100" : ["1.5.1"]},
-            "tensorflow-gpu" : {
-                        "90": ["1.12.0",],
-                        "100": ["1.14.0", "1.15.2", "2.0.1", "2.1.0"]},
-            "torch" : {
-                        "90": ["http://download.pytorch.org/whl/{cuda_v}/torch-1.1.0-{0}.whl"],
-                        "100": ["http://download.pytorch.org/whl/{cuda_v}/torch-1.4.0+{cuda_v}-{0}.whl"]},
-            "torchvision" : {
-                        "90": ["https://download.pytorch.org/whl/{cuda_v}/torchvision-0.3.0-{0}.whl"],
-                        "100": ["https://download.pytorch.org/whl/{cuda_v}/torchvision-0.5.0+{cuda_v}-{0}.whl"]},
-            "paddle" : {
-                        "90": ["https://paddle-wheel.bj.bcebos.com/gcc54/latest-gpu-cuda9-cudnn7-openblas/paddlepaddle_gpu-latest-{0}.whl"],
-                        "100": ["https://paddle-wheel.bj.bcebos.com/gcc54/latest-gpu-cuda10-cudnn7-openblas/paddlepaddle_gpu-latest-{0}.whl"]},
-            }
+PYTHON_VERSION = ".".join([str(x) for x in sys.version_info[0:2]])
+
+class PckgVer():
+    """Class that holds a version string accompanied with maximum and minimum python version that
+        this version should support. If python falls beyond version bounds it evaluates to the empty string
+
+        Parameters
+        ----------
+        `ver`: str
+            Version that is housed by object of this class
+        `python_max_ver` : str, optional, default = None
+            Maximum python version supported by this package. If empty there is no upper bound
+        `python_min_ver`: str, optional, default = None
+            Mimimum python version supported by this package. If empty there is no lower bound
+    """
+    def __init__(self, ver, python_min_ver=None, python_max_ver=None):
+        self.ver = ver
+        self.python_min_ver = python_min_ver
+        self.python_max_ver = python_max_ver
+
+    def __bool__(self):
+        return (not self.python_min_ver or parse(PYTHON_VERSION) >= parse(self.python_min_ver)) and \
+           (not self.python_max_ver or parse(PYTHON_VERSION) <= parse(self.python_max_ver))
+
+    def __repr__(self):
+        if self:
+            return self.ver
+        else:
+            return ""
+
+class BasePackage():
+    """Class describing basic methods that package should provide
+
+        Parameters
+        ----------
+        `key`: str
+            Name this package should be queried for
+        `versions`: str
+            List of versions this package is available for
+        `name`: str, , optional, default = None
+            Name of this package used during installation. If empty it is the same as the key
+    """
+    def __init__(self, key, versions, name=None):
+        self.key = key
+        if not name:
+            name = key
+        self.name = name
+        self.versions = versions
+
+    """Clams index to range 0 - num_of_packages - 1
```

Comment:
```suggestion
    """Clamps index to range 0 - num_of_packages - 1
```

And I think the docstrings should be provided under function signature in Python.

Reply:
Done